{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total coauthors:352\n",
      "Nayfach et al., \"A genomic catalog of Earth’s microbiomes,\" Nature biotechnology, 2021.\n",
      "Seaver et al., \"The ModelSEED Biochemistry Database for the integration of metabolic annotations and the reconstruction, comparison and analysis of metabolic models for plants, fungi and microbes,\" Nucleic acids research, 2021.\n",
      "Rubinstein et al., \"ORT: A workflow linking genome-scale metabolic models with reactive transport codes,\" bioRxiv, 2021.\n",
      "Hanson et al., \"The number of catalytic cycles in an enzyme’s lifetime and why it matters to metabolic engineering,\" Proceedings of the National Academy of Sciences, 2021.\n",
      "Joshi et al., \"Quantum Mechanical Methods Predict Accurate Thermodynamics of Biochemical Reactions,\" ACS omega, 2021.\n",
      "Lagoa et al., \"TranSyT, the Transport Systems Tracker,\" bioRxiv, 2021.\n",
      "van der Lelie et al., \"Rationally designed bacterial consortia to treat chronic immune-mediated colitis and restore intestinal homeostasis,\" Nature communications, 2021.\n",
      "Qiu et al., \"ELRUNA: Elimination Rule-based Network Alignment,\" Journal of Experimental Algorithmics (JEA), 2021.\n",
      "Lelie et al., \"Rationally designed bacterial consortia to treat chronic immune-mediated colitis and restore intestinal homeostasis,\" Nature Communications, 2021.\n",
      "Khazaei et al., \"Metabolic multistability and hysteresis in a model aerobe-anaerobe microbiome community,\" Science advances, 2020.\n",
      "Lieven et al., \"MEMOTE for standardized genome-scale metabolic model testing,\" Nature biotechnology, 2020.\n",
      "Tivendale et al., \"Enzymes as parts in need of replacement–and how to extend their working life,\" Trends in Plant Science, 2020.\n",
      "Seaver et al., \"The ModelSEED Database for the integration of metabolic annotations and the reconstruction, comparison, and analysis of metabolic models for plants, fungi, and microbes,\" bioRxiv, 2020.\n",
      "Gupta et al., \"Draft Genome Sequence of 2-Methylpyridine-, 2-Ethylpyridine-, and 2-Hydroxypyridine-Degrading Arthrobacter sp. Strain ATCC 49987,\" Microbiology resource announcements, 2020.\n",
      "Zhao et al., \"Chemical composition of material extractives influences microbial growth and dynamics on wetted wood materials,\" Scientific reports, 2020.\n",
      "Gupta et al., \"Draft Genome Sequence of Pseudarthrobacter sp. Strain ATCC 49442 (Formerly Micrococcus luteus), a Pyridine-Degrading Bacterium,\" Microbiology resource announcements, 2020.\n",
      "Heinken et al., \"AGORA2: large scale reconstruction of the microbiome highlights wide-spread drug-metabolising capacities,\" BioRxiv, 2020.\n",
      "Song et al., \"Machine Learning-based Metabolic Network Modeling for Omics-integrated Biogeochemical and Reactive Transport Simulations,\" AGU Fall Meeting Abstracts, 2020.\n",
      "Kessell et al., \"Metabolic Network Modeling and Metabolomics Integration for Comparative Analysis of Biogeochemical Reactions in Multiple River Systems,\" AGU Fall Meeting Abstracts, 2020.\n",
      "Shaffer et al., \"Distributed Metagenomic Sampling of River Corridors Identifies Shared Microbial Functional Traits Underpinning Transferable Hydrobiogeochemical Processes,\" AGU Fall Meeting Abstracts, 2020.\n",
      "Lax et al., \"Microbial and metabolic succession on common building materials under high humidity conditions,\" Nature communications, 2019.\n",
      "Jeffryes et al., \"Metabolic in silico network expansions to predict and exploit enzyme promiscuity,\" Microbial Metabolic Engineering, 2019.\n",
      "Gupta et al., \"Draft genome sequence of Enterobacter sp. strain A8, a carbazole-degrading bacterium,\" Microbiology resource announcements, 2019.\n",
      "Gupta et al., \"Draft genome sequence of Rhodococcus sp. strain ATCC 49988, a quinoline-degrading bacterium,\" Microbiology resource announcements, 2019.\n",
      "Qiu et al., \"Network Alignment by Propagating Reliable Similarities,\" CoRR, 2019.\n",
      "Edirisinghe et al., \"Reconstruction and analysis of central metabolism in microbes,\" Metabolic Network Reconstruction and Modeling, 2018.\n",
      "Song et al., \"Metabolic network modeling for computer-aided design of microbial interactions,\" Emerg. Areas Bioeng, 2018.\n",
      "Huang et al., \"A plastidial pantoate transporter with a potential role in pantothenate synthesis,\" Biochemical Journal, 2018.\n",
      "Jeffryes et al., \"A pathway for every product? Tools to discover and design plant metabolism,\" Plant Science, 2018.\n",
      "Seaver et al., \"PlantSEED enables automated annotation and reconstruction of plant primary metabolism with improved compartmentalization and comparative consistency,\" The Plant Journal, 2018.\n",
      "Lieven et al., \"Memote: A community driven effort towards a standardized genome-scale metabolic model test suite,\" BioRxiv, 2018.\n",
      "Arkin et al., \"KBase: the United States department of energy systems biology knowledgebase,\" Nature biotechnology, 2018.\n",
      "Faria et al., \"Methods for automated genome-scale metabolic model reconstruction,\" Biochemical Society Transactions, 2018.\n",
      "Li et al., \"Implementing and evaluating a Gaussian mixture framework for identifying gene function from TnSeq data,\" BIOCOMPUTING 2019: Proceedings of the Pacific Symposium, 2018.\n",
      "Cardona et al., \"Environmental sources of bacteria differentially influence host-associated microbial dynamics,\" MSystems, 2018.\n",
      "Marshall et al., \"Metabolic reconstruction and modeling microbial electrosynthesis,\" Scientific reports, 2017.\n",
      "Wattam et al., \"Improvements to PATRIC, the all-bacterial bioinformatics database and analysis resource center,\" Nucleic acids research, 2017.\n",
      "van der Lelie et al., \"The microbiome as a source of new enterprises and job creation: considering clinical faecal and synthetic microbiome transplants and therapeutic regulation,\" Microbial biotechnology, 2017.\n",
      "Wu et al., \"A novel signal transduction protein: Combination of solute binding and tandem PAS‐like sensor domains in one polypeptide chain,\" Protein Science, 2017.\n",
      "Henry et al., \"Generation and validation of the iKp1289 metabolic model for Klebsiella pneumoniae KPPR1,\" The Journal of infectious diseases, 2017.\n",
      "Juárez-Vázquez et al., \"Evolution of substrate specificity in a retained enzyme driven by gene loss,\" Elife, 2017.\n",
      "Sun et al., \"Metabolite damage and repair in metabolic engineering design,\" Metabolic engineering, 2017.\n",
      "Weisenhorn et al., \"Comparative'Omics and Modeling Microbial Community Interactions Using the DOE Systems Biology Knowledgebase (KBase),\" 2017 ESA Annual Meeting (August 6--11), 2017.\n",
      "Henry et al., \"Development and Application of Integrated Pipeline for the Modeling and Analysis of Microbial Communities in the Doe Systems Biology Knowledgebase,\" 2017 AIChE Annual Meeting, 2017.\n",
      "Hanson et al., \"Metabolite damage and metabolite damage control in plants,\" Annual review of plant biology, 2016.\n",
      "Oberhardt et al., \"Systems-wide prediction of enzyme promiscuity reveals a new underground alternative route for pyridoxal 5’-phosphate production in e. coli,\" PLoS computational biology, 2016.\n",
      "Verduzco-Castro et al., \"Co-occurrence of analogous enzymes determines evolution of a novel (βα) 8-isomerase sub-family after non-conserved mutations in flexible loop,\" Biochemical Journal, 2016.\n",
      "Faria et al., \"Reconstruction of the regulatory network for Bacillus subtilis and reconciliation with gene expression data,\" Frontiers in microbiology, 2016.\n",
      "Thiaville et al., \"Experimental and metabolic modeling evidence for a folate-cleaving side-activity of ketopantoate hydroxymethyltransferase (PanB),\" Frontiers in microbiology, 2016.\n",
      "Henry et al., \"Microbial community metabolic modeling: a community data‐driven network reconstruction,\" Journal of cellular physiology, 2016.\n",
      "Ozer et al., \"Draft genome sequence of a multidrug-resistant Klebsiella quasipneumoniae subsp. similipneumoniae isolate from a clinical source,\" Genome announcements, 2016.\n",
      "Lerma-Ortiz et al., \"‘Nothing of chemistry disappears in biology’: the Top 30 damage-prone endogenous metabolites,\" Biochemical Society Transactions, 2016.\n",
      "Faria et al., \"Constructing and analyzing metabolic flux models of microbial communities,\" Hydrocarbon and Lipid Microbiology Protocols, 2016.\n",
      "Henry et al., \"Systematic identification and analysis of frequent gene fusion events in metabolic pathways,\" BMC genomics, 2016.\n",
      "Edirisinghe et al., \"Modeling central metabolism and energy biosynthesis across microbial life,\" Bmc Genomics, 2016.\n",
      "Faria et al., \"Computing and applying atomic regulons to understand gene expression and regulation,\" Frontiers in microbiology, 2016.\n",
      "Gillum et al., \"Commentary Article,\" BIOCHEMICAL JOURNAL, 2016.\n",
      "Verduzco-Castro et al., \"Co-occurrence of analogous enzymes determines evolution of a novel () 8-isomerase sub-family after non-conserved mutations in flexible loop,\" Biochemical Journal, 2016.\n",
      "Cuevas et al., \"From DNA to FBA: how to build your own genome-scale metabolic model,\" Frontiers in microbiology, 2016.\n",
      "Cardona et al., \"Network-based metabolic analysis and microbial community modeling,\" Current opinion in microbiology, 2016.\n",
      "Cuevas et al., \"Elucidating genomic gaps using phenotypic profiles,\" F1000Research, 2016.\n",
      "Faria et al., \"Enabling comparative modeling of closely related genomes: example genus Brucella,\" 3 Biotech, 2015.\n",
      "Plata et al., \"Long-term phenotypic evolution of bacteria,\" Nature, 2015.\n",
      "Seaver et al., \"Improved evidence-based genome-scale metabolic models for maize leaf, embryo, and endosperm,\" Frontiers in plant science, 2015.\n",
      "Jeffryes et al., \"MINEs: open access databases of computationally predicted enzyme promiscuity products for untargeted metabolomics,\" Journal of cheminformatics, 2015.\n",
      "Frelin et al., \"A directed-overflow and damage-control N-glycosidase in riboflavin biosynthesis,\" Biochemical Journal, 2015.\n",
      "Gilbert et al., \"Predicting ecosystem emergent properties at multiple scales,\" Environmental microbiology reports, 2015.\n",
      "Faria et al., \"Genome-scale bacterial transcriptional regulatory networks: reconstruction and integrated analysis with metabolic models,\" Briefings in bioinformatics, 2014.\n",
      "Fagen et al., \"Comparative genomics of cultured and uncultured strains suggests genes essential for free-living growth of Liberibacter,\" PLoS One, 2014.\n",
      "Zarecki et al., \"Maximal sum of metabolic exchange fluxes outperforms biomass yield as a predictor of growth rate of microorganisms,\" PloS one, 2014.\n",
      "Seaver et al., \"High-throughput comparison, functional annotation, and metabolic modeling of plant genomes using the PlantSEED resource,\" Proceedings of the National Academy of Sciences, 2014.\n",
      "Colasanti et al., \"Tapping the Wealth of Microbial Data in High-Throughput Metabolic Model Reconstruction,\" Metabolic Flux Analysis, 2014.\n",
      "Benedict et al., \"Likelihood-based gene annotations for gap filling and quality assessment in genome-scale metabolic models,\" PLoS computational biology, 2014.\n",
      "Tanaka et al., \"Building the repertoire of dispensable chromosome regions in Bacillus subtilis entails major refinement of cognate large-scale metabolic model,\" Nucleic acids research, 2013.\n",
      "Devoid et al., \"Automated genome annotation and metabolic model reconstruction in the SEED and Model SEED,\" Systems Metabolic Engineering, 2013.\n",
      "Krumholz et al., \"Genome-wide metabolic network reconstruction of the picoalga Ostreococcus,\" Journal of experimental botany, 2012.\n",
      "Seaver et al., \"Frontiers in metabolic reconstruction and modeling of plant genomes,\" Journal of experimental botany, 2012.\n",
      "Gerdes et al., \"Plant B vitamin pathways and their compartmentation: a guide for the perplexed,\" Journal of experimental botany, 2012.\n",
      "Aziz et al., \"SEED servers: high-performance access to the SEED genomes, annotations, and metabolic models,\" PloS one, 2012.\n",
      "Henry et al., \"Connecting genotype to phenotype in the era of high-throughput sequencing,\" Biochimica et Biophysica Acta (BBA)-General Subjects, 2011.\n",
      "Larsen et al., \"Predicted Relative Metabolomic Turnover (PRMT): determining metabolic turnover from a coastal marine metagenomic dataset,\" Microbial informatics and experimentation, 2011.\n",
      "Freilich et al., \"Competitive and cooperative metabolic interactions in bacterial communities,\" Nature communications, 2011.\n",
      "Larsen et al., \"Relative metabolomic flux: Predicting changes in the environmental metabolome from the metagenome,\" Microbial Informatics and Experimentation, 2011.\n",
      "Henry et al., \"High-throughput generation, optimization and analysis of genome-scale metabolic models,\" Nature biotechnology, 2010.\n",
      "Henry et al., \"Discovery and analysis of novel metabolic pathways for the biosynthesis of industrial chemicals: 3‐hydroxypropanoate,\" Biotechnology and bioengineering, 2010.\n",
      "Henry et al., \"Building the blueprint of life,\" Biotechnology journal, 2010.\n",
      "Faria et al., \"Analysis of the effect of reversibility constraints on the predictions of genome-scale metabolic models,\" Advances in Bioinformatics, 2010.\n",
      "Henry et al., \"iBsu1103: a new genome-scale metabolic model of Bacillus subtilis based on SEED annotations,\" Genome Biol, 2009.\n",
      "Henry et al., \"Application of high-performance computing to the reconstruction, analysis, and optimization of genome-scale metabolic models,\" Journal of Physics: Conference Series, 2009.\n",
      "Broadbelt et al., \"Discovery of novel routes for the production of fuels and chemicals,\" Design for Energy and the Environment, 2009.\n",
      "Jankowski et al., \"Group contribution method for thermodynamic analysis of complex metabolic networks,\" Biophysical journal, 2008.\n",
      "Feist et al., \"A genome‐scale metabolic reconstruction for Escherichia coli K‐12 MG1655 that accounts for 1260 ORFs and thermodynamic information,\" Molecular systems biology, 2007.\n",
      "Henry et al., \"Thermodynamics-based metabolic flux analysis,\" Biophysical journal, 2007.\n",
      "Henry et al., \"Computational thermodynamic and biosynthetic analysis of genome-scale metabolic models,\" , 2007.\n",
      "Henry et al., \"Genome-scale thermodynamic analysis of Escherichia coli metabolism,\" Biophysical journal, 2006.\n",
      "Hatzimanikatis et al., \"Exploring the diversity of complex metabolic networks,\" Bioinformatics, 2005.\n",
      "Li et al., \"Computational discovery of biochemical routes to specialty chemicals,\" Chemical engineering science, 2004.\n",
      "Wilkens et al., \"How to scale-up mixing processes in non-Newtonian fluids,\" Chem Eng Prog, 2003.\n",
      "Kutt et al., \"Automatic generation of genome-scale metabolism and expression models for bacteria,\" , .\n",
      "Sedova et al., \"Genome to Structure: From Multiple Sequence Alignment to Virtual Ligand Screening Using Co-Evolutionary Protein-Residue Contact-Prediction in KBase,\" , .\n",
      "Cuevas et al., \"High-throughput phenotype profiling for bacterial flux-balance model optimization,\" Proteins, .\n",
      "Arkin et al., \"New pipeline for multi-omics data integration and discovery in KBase to identify the mechanism driving metabolic/OTU dynamics in a microbiome,\" , .\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "data = {}\n",
    "author_hash = {}\n",
    "first_only = True\n",
    "with open('citations.csv', newline='') as csvfile:\n",
    "    citations = csv.reader(csvfile)\n",
    "    header = None\n",
    "    for row in citations:\n",
    "        citation = \"\"\n",
    "        if header == None:\n",
    "            header = row\n",
    "        else:\n",
    "            authors = row[0].split(\"; \")\n",
    "            for author in authors:\n",
    "                if len(author) >= 2:\n",
    "                    if len(citation) > 0:\n",
    "                        if not first_only:\n",
    "                            citation += \", \"\n",
    "                    array = author.split(\", \")\n",
    "                    if len(array) > 1:\n",
    "                        name = array[0]+\", \"\n",
    "                        newarray = array[1].split(\" \")\n",
    "                        for item in newarray:\n",
    "                            name += item[0]+\".\"\n",
    "                        author_hash[name] = 1\n",
    "                        if first_only and len(citation) == 0:\n",
    "                            citation += array[0]\n",
    "                            if len(authors) > 1:\n",
    "                                citation += \" et al.\"\n",
    "                        elif not first_only:\n",
    "                            citation += name\n",
    "            row[3] = \"\"\n",
    "            row[4] = \"\"\n",
    "            if len(row[4]) == 0:\n",
    "                if len(row[3]) == 0:\n",
    "                    citation += ', \"'+row[1]+',\" '+row[2]+\", \"+row[6]+\".\"\n",
    "                else:\n",
    "                    citation += ', \"'+row[1]+',\" '+row[2]+\", \"+row[3]+\", \"+row[6]+\".\"\n",
    "            else:\n",
    "                citation += ', \"'+row[1]+',\" '+row[2]+\", \"+row[3]+\"(\"+row[4]+\"), \"+row[6]+\".\"\n",
    "            if row[6] not in data:\n",
    "                data[row[6]] = []\n",
    "            data[row[6]].append(citation)\n",
    "            keylist = []\n",
    "            for year in data.keys():\n",
    "                keylist.append(year)\n",
    "            keylist.sort(reverse = True)\n",
    "\n",
    "print(\"Total coauthors:\"+str(len(author_hash)))\n",
    "for year in keylist:\n",
    "    for citation in data[year]:\n",
    "        print(citation)\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# constraints\n",
    "# solution\n",
    "# visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def constraints(modelInfo,mediaInfo,kinetic_coeff,abundances=None):\n",
    "#     kbase_api = cobrakbase.KBaseAPI()\n",
    "#     model = kbase_api.get_from_ws(modelInfo[0],modelInfo[1])\n",
    "#     model.solver = 'optlang-cplex'\n",
    "\n",
    "#     #Applying uptake limit to total carbon uptake by model\n",
    "#     eup = ElementUptakePkg(model)\n",
    "#     eup.build_package({\"C\":60})\n",
    "#     #Applying constraints based on KBase media\n",
    "#     media = kbase_api.get_from_ws(mediaInfo[0],mediaInfo[1])\n",
    "#     kmp = KBaseMediaPkg(model)\n",
    "#     kmp.build_package(media)\n",
    "#     #Applying community model kinetic constraints\n",
    "#     ckp = CommKineticPkg(model)\n",
    "#     ckp.build_package(kinetic_coeff,abundances) # <- SPECIES ABUNDANCE BUG?\n",
    "\n",
    "#     biomass_objective = model.problem.Objective(\n",
    "#         1 * model.reactions.bio1.flux_expression,\n",
    "#         direction='max')\n",
    "#     model.objective = biomass_objective\n",
    "#     return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def communityFBA(modelInfo,mediaInfo,kinetic_coeff,abundances=None):\n",
    "    kbase_api = cobrakbase.KBaseAPI()\n",
    "    model = kbase_api.get_from_ws(modelInfo[0],modelInfo[1])\n",
    "    model.solver = 'optlang-cplex'\n",
    "\n",
    "    #Applying uptake limit to total carbon uptake by model\n",
    "    eup = ElementUptakePkg(model)\n",
    "    eup.build_package({\"C\":60})\n",
    "    #Applying constraints based on KBase media\n",
    "    media = kbase_api.get_from_ws(mediaInfo[0],mediaInfo[1])\n",
    "    kmp = KBaseMediaPkg(model)\n",
    "    kmp.build_package(media)\n",
    "    #Applying community model kinetic constraints\n",
    "    ckp = CommKineticPkg(model)\n",
    "    ckp.build_package(kinetic_coeff,abundances)\n",
    "\n",
    "    biomass_objective = model.problem.Objective(\n",
    "        1 * model.reactions.bio1.flux_expression,\n",
    "        direction='max')\n",
    "    model.objective = biomass_objective\n",
    "\n",
    "    # FBA\n",
    "    sol=model.optimize()\n",
    "    \n",
    "    # pFBA\n",
    "#     sol = cobra.flux_analysis.pfba(model)\n",
    "    \n",
    "    #print(model.summary())\n",
    "\n",
    "    # Compute exchanges\n",
    "    metabolite_uptake = {}\n",
    "    speciesNum = []\n",
    "    for rxn in model.reactions:\n",
    "        # obtain species\n",
    "        ind = rxn.id.rfind('_')\n",
    "        if (rxn.id[ind + 1] == 'c' or rxn.id[ind + 1] == 'p') and rxn.id[ind + 2: ] != '0':\n",
    "            species = rxn.id[ind + 2: ]\n",
    "            speciesNum.append(species)\n",
    "\n",
    "            for metabolite in rxn.metabolites:\n",
    "                if metabolite.compartment == \"e0\":\n",
    "                    rate = 0\n",
    "                    flux = sol.fluxes[rxn.id]\n",
    "                    if flux != 0:\n",
    "                        rate += rxn.metabolites[metabolite]*flux\n",
    "                    if rate != 0:\n",
    "                        metabolite_uptake[(metabolite.id,species)] = rate\n",
    "\n",
    "    speciesNum = np.unique(speciesNum)\n",
    "    numOfSpecies = len(speciesNum)\n",
    "\n",
    "    # Cross feeding\n",
    "    external_metabolites = []\n",
    "    for rxn in model.reactions:\n",
    "        for metabolite in rxn.metabolites:\n",
    "            if metabolite.compartment == \"e0\":\n",
    "                external_metabolites.append(metabolite.id)\n",
    "\n",
    "    cross_all = []\n",
    "    for metabolite in external_metabolites:\n",
    "        r_out = {x: r for (m, x), r in metabolite_uptake.items() if m == metabolite and r > 0}\n",
    "        r_in = {x: -r for (m, x), r in metabolite_uptake.items() if m == metabolite and -r > 0}\n",
    "        total_in = sum(r_in.values())\n",
    "        total_out = sum(r_out.values())\n",
    "        total = max(total_in, total_out)\n",
    "        if total_in > total_out:\n",
    "            r_out[None] = total_in - total_out\n",
    "        if total_out > total_in:\n",
    "            r_in[None] = total_out - total_in\n",
    "\n",
    "        cross = [(o1, o2, metabolite, r1 * r2 / total) for o1, r1 in r_out.items() for o2, r2 in r_in.items()]\n",
    "        cross_all.extend(cross) # [donor, receiver, compound, rate]\n",
    "\n",
    "#     for i in cross_all:\n",
    "#         print(i)\n",
    "\n",
    "    production = [[0] * numOfSpecies for i in range(numOfSpecies)] #production[donorSpecies][receiverSpecies]\n",
    "    consumption = [[0] * numOfSpecies for i in range(numOfSpecies)]\n",
    "\n",
    "    for i in cross_all:\n",
    "        if i[0] is not None and i[1] is not None:\n",
    "            production[int(i[0])-1][int(i[1])-1] += i[3]\n",
    "            consumption[int(i[1])-1][int(i[0])-1] += i[3]\n",
    "\n",
    "    # for i in production:\n",
    "    #     print(i)\n",
    "    # print(' ')\n",
    "    # for i in consumption:\n",
    "    #     print(i)\n",
    "    \n",
    "    G=nx.Graph()\n",
    "    for i in speciesNum:\n",
    "        G.add_node(i)\n",
    "    for k in combinations(speciesNum, 2):\n",
    "        sp1 = int(k[0])-1\n",
    "        sp2 = int(k[1])-1\n",
    "        if production[sp1][sp2] > consumption[sp2][sp1]:\n",
    "            interaction_flux = production[sp1][sp2]\n",
    "        else:\n",
    "            interaction_flux = consumption[sp2][sp1]\n",
    "        G.add_edge(k[0],k[1],weight=interaction_flux)\n",
    "\n",
    "    pos=nx.circular_layout(G)\n",
    "    nx.draw_networkx(G,pos)\n",
    "    labels = nx.get_edge_attributes(G,'weight')\n",
    "    nx.draw_networkx_edge_labels(G,pos,edge_labels=labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cobrakbase' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-896c5c627033>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmodelInfo_2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"CMM_iAH991V2_iML1515.kb\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m40576\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mmediaInfo_2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"Btheta_Ecoli_minimal_media\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m40576\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mcommunityFBA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodelInfo_2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmediaInfo_2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2100\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m\"1\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m0.5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"2\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m0.5\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-4-b5015bc5fcd2>\u001b[0m in \u001b[0;36mcommunityFBA\u001b[0;34m(modelInfo, mediaInfo, kinetic_coeff, abundances)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mcommunityFBA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodelInfo\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmediaInfo\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mkinetic_coeff\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mabundances\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mkbase_api\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcobrakbase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mKBaseAPI\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkbase_api\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_from_ws\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodelInfo\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmodelInfo\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msolver\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'optlang-cplex'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'cobrakbase' is not defined"
     ]
    }
   ],
   "source": [
    "modelInfo_2 = [\"CMM_iAH991V2_iML1515.kb\",40576]\n",
    "mediaInfo_2 = [\"Btheta_Ecoli_minimal_media\",40576]\n",
    "communityFBA(modelInfo_2,mediaInfo_2,2100,{\"1\":0.5,\"2\":0.5})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changing the kinetic coefficient\n",
    "communityFBA(modelInfo_2,mediaInfo_2,2300,{\"1\":0.5,\"2\":0.5})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changing the species abundances\n",
    "communityFBA(modelInfo_2,mediaInfo_2,2100,{\"1\":0.1,\"2\":0.9})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modelInfo_3 = [\"electro.community.edited.v2\",93204]\n",
    "mediaInfo_3 = [\"CO2_minimal\",93204]\n",
    "communityFBA(modelInfo_3,mediaInfo_3,2100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changing the kinetic coefficient\n",
    "communityFBA(modelInfo_3,mediaInfo_3,2300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changing the species abundances\n",
    "communityFBA(modelInfo_3,mediaInfo_3,2100,abundances={\"1\":0.1,\"2\":0.1,\"3\":0.8})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# modelInfo_7 = [\"Hot_Lake_seven.mdl.v2\",94153]\n",
    "# mediaInfo_7 = [\"HotLakeMedia\",94153]\n",
    "# communityFBA(modelInfo_7,mediaInfo_7,2100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
