{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python version 3.7.6\n",
      "cobrakbase 0.2.8\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-e0ecfa9ff4c1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mcobra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflux_analysis\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mflux_variability_analysis\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mmodelseedpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhelpers\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mget_template\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0msklearn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetrics\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mr2_score\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_line_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'matplotlib'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'inline'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mmatplotlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpyplot\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/sklearn/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     80\u001b[0m     \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m_distributor_init\u001b[0m  \u001b[0;31m# noqa: F401\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     81\u001b[0m     \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m__check_build\u001b[0m  \u001b[0;31m# noqa: F401\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 82\u001b[0;31m     \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mbase\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mclone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     83\u001b[0m     \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_show_versions\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mshow_versions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/sklearn/base.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m__version__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m_IS_32BIT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m _DEFAULT_TAGS = {\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/sklearn/utils/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mscipy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msparse\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0missparse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mmurmurhash\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mmurmurhash3_32\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mclass_weight\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mcompute_class_weight\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcompute_sample_weight\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m_joblib\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/importlib/_bootstrap.py\u001b[0m in \u001b[0;36mparent\u001b[0;34m(self)\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import platform\n",
    "print(\"python version \" + platform.python_version())\n",
    "import sys\n",
    "import json\n",
    "import cobra\n",
    "import cplex\n",
    "import re\n",
    "import os\n",
    "from os.path import exists\n",
    "import logging\n",
    "from configparser import ConfigParser\n",
    "config = ConfigParser()\n",
    "config.read(\"config.cfg\")\n",
    "paths = config.get(\"script\", \"syspaths\").split(\";\")\n",
    "for path in paths:\n",
    "    sys.path.append(path)\n",
    "import cobrakbase\n",
    "from escher import Builder\n",
    "from optlang.symbolics import Zero, add\n",
    "from modelseedpy import MSPackageManager, MSMedia, MSModelUtil, MSGapfill, FBAHelper, MSGrowthPhenotypes, MSModelUtil, MSATPCorrection\n",
    "from cobrakbase.core.kbasefba.newmodeltemplate_builder import NewModelTemplateBuilder\n",
    "from annotation_ontology_api.annotation_ontology_apiServiceClient import annotation_ontology_api\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "from modelseedpy.helpers import get_template\n",
    "from sklearn.metrics import r2_score\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from IPython.display import HTML\n",
    "print(\"Required modules loaded\")\n",
    "kbase_api = cobrakbase.KBaseAPI()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes = kbase_api.list_objects(80735, object_type=\"KBaseGenomes.Genome\", include_metadata=True)\n",
    "genome_hash = {}\n",
    "for genome in genomes:\n",
    "    genome_hash[genome[1]] = {\n",
    "        \"genes\":genome[10][\"Number of Protein Encoding Genes\"],\n",
    "        \"species\":genome[10][\"Name\"],\n",
    "        \"taxonomy\":genome[10][\"Taxonomy\"]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('AuxoProblemModels.txt')\n",
    "genome_list = f.read().split(\"\\n\")\n",
    "filter_hash = {}\n",
    "for item in genome_list:\n",
    "    filter_hash[item] = 1\n",
    "#mdlws = 114104#Glucose minimal media workspace\n",
    "#Loading cached data\n",
    "#f = open('GMMHash.json')\n",
    "#genome_hash = json.load(f)\n",
    "mdlws = 114650#Auxotrophy media workspace\n",
    "models = kbase_api.list_objects(mdlws, object_type=\"KBaseFBA.FBAModel\", include_metadata=True)\n",
    "for item in models:\n",
    "    genomeid = item[1][0:-4]\n",
    "    if genomeid in filter_hash and \"reactions\" not in genome_hash[genomeid]:\n",
    "        print(genomeid)\n",
    "        model = kbase_api.get_from_ws(item[1],mdlws)\n",
    "        pkgmgr = MSPackageManager.get_pkg_mgr(model)\n",
    "        pkgmgr.getpkg(\"KBaseMediaPkg\").build_package(None)\n",
    "        fva = flux_variability_analysis(model,model.reactions,fraction_of_optimum=0.1)\n",
    "        gfcount = 0\n",
    "        blocked = 0\n",
    "        for reaction in model.reactions:\n",
    "            if len(reaction.genes) == 0:\n",
    "                gfcount += 1\n",
    "            if fva[\"maximum\"][reaction.id] == 0 and fva[\"minimum\"][reaction.id] == 0:\n",
    "                blocked += 1\n",
    "        gfcount += -22\n",
    "        genome_hash[genomeid][\"reactions\"] = len(model.reactions)\n",
    "        genome_hash[genomeid][\"mdlgenes\"] = len(model.genes)\n",
    "        genome_hash[genomeid][\"gfreactions\"] = gfcount\n",
    "        genome_hash[genomeid][\"blocked\"] = blocked\n",
    "        with open('problem_hash.json', 'w') as outfile:\n",
    "            json.dump(genome_hash, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'kbase_api' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-1e730ff0ab95>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0muniversals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtemplate\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkbase_api\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_from_ws\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'GramPosModelTemplateV4'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m12998\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mreaction\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtemplate\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreactions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mreaction\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"universal\"\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mreaction\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"spontaneous\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0muniversals\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mreaction\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"0\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'kbase_api' is not defined"
     ]
    }
   ],
   "source": [
    "universals = {}\n",
    "template = kbase_api.get_from_ws('GramPosModelTemplateV4', 12998)\n",
    "for reaction in template.reactions:\n",
    "    if reaction.type == \"universal\" or reaction.type == \"spontaneous\":\n",
    "        universals[reaction.id+\"0\"] = 1\n",
    "auxotrans = [\n",
    "   \"rxn00068\",\"rxn09693\",\"rxn05663\",\"rxn05301\",\"rxn05306\",\"rxn05669\",\"rxn05244\",\"rxn05243\",\"rxn05496\",\"rxn05217\",\"rxn05508\",\"rxn05307\",\"rxn05300\",\"rxn05582\",\"rxn09672\",\"rxn09696\",\"rxn09690\",\"rxn05297\",\"rxn09678\",\"rxn05305\",\"rxn05299\",\"rxn05303\",\"rxn05638\",\"rxn05687\",\"rxn05652\",\"rxn12666\",\"rxn08192\",\"rxn10147\",\"rxn05310\",\"rxn05645\",\"rxn05308\",\"rxn05255\",\"rxn09657\",\"rxn05148\" \n",
    "]\n",
    "for rxn in auxotrans:\n",
    "    universals[rxn+\"_c0\"] = 1\n",
    "f = open('auxo_hash.json')\n",
    "genome_hash = json.load(f)\n",
    "mdlws = 114104#Minimal media workspace\n",
    "#Creating auxotrophy media\n",
    "compound_sets = [\n",
    "    [\"cpd00065\",\"cpd00069\",\"cpd00066\",\"cpd00393\"],#AAA+folate\n",
    "    [\"cpd00118\",\"cpd00264\"],#put,sper\n",
    "    [\"cpd00028\",\"cpd00557\",\"cpd00635\"],#heme,cbl\n",
    "    [\"cpd00033\",\"cpd00054\"],#gly,ser\n",
    "    [\"cpd00039\",\"cpd00161\"],#lys,thre\n",
    "    [\"cpd00107\",\"cpd00156\",\"cpd00322\"],#leu,val,isoleu\n",
    "    [\"cpd00065\"],\n",
    "    [\"cpd00069\"],\n",
    "    [\"cpd00066\"],\n",
    "    [\"cpd00156\"],\n",
    "    [\"cpd00322\"],\n",
    "    [\"cpd00107\"],\n",
    "    [\"cpd00132\"],\n",
    "    [\"cpd00054\"],\n",
    "    [\"cpd00161\"],\n",
    "    [\"cpd00033\"],\n",
    "    [\"cpd00060\"],\n",
    "    [\"cpd00084\"],\n",
    "    [\"cpd00039\"],\n",
    "    [\"cpd00119\"],\n",
    "    [\"cpd00051\"],\n",
    "    [\"cpd00129\"],\n",
    "    [\"cpd00118\"],\n",
    "    [\"cpd00264\"],\n",
    "    [\"cpd00028\"],\n",
    "    [\"cpd00557\"],\n",
    "    [\"cpd00635\"],\n",
    "    [\"cpd00218\"],\n",
    "    [\"cpd00220\"],\n",
    "    [\"cpd00644\"],\n",
    "    [\"cpd00393\"],\n",
    "    [\"cpd00305\"],\n",
    "    [\"cpd00104\"],\n",
    "    [\"cpd00215\"]\n",
    "]\n",
    "gmmedia = kbase_api.get_from_ws(\"Carbon-D-Glucose\",\"KBaseMedia\")\n",
    "input_dictionary = {}\n",
    "for cpdset in compound_sets:\n",
    "    for cpd in cpdset:\n",
    "        input_dictionary[cpd] = 1\n",
    "auxomedia = MSMedia.from_dict(input_dictionary)\n",
    "auxomedia.merge(gmmedia)\n",
    "#Iterating over models and checking each for auxotrophy analysis\n",
    "models = kbase_api.list_objects(mdlws, object_type=\"KBaseFBA.FBAModel\", include_metadata=True)\n",
    "for item in models:\n",
    "    genomeid = item[1][0:-4]\n",
    "    if genomeid in genome_hash and \"reactions\" not in genome_hash[genomeid]:\n",
    "        print(genomeid)\n",
    "        model = kbase_api.get_from_ws(item[1],mdlws)\n",
    "        mdlutl = MSModelUtil(model)\n",
    "        mdlutl.add_missing_exchanges(auxomedia)\n",
    "        genome_hash[genomeid][\"auxo\"] = {}\n",
    "        model.reactions.bio1.lower_bound = 0.1\n",
    "        gfreactions = []\n",
    "        model.objective = model.problem.Objective(\n",
    "            Zero,\n",
    "            direction=\"min\")\n",
    "        obj_coef = dict()\n",
    "        for reaction in model.reactions:\n",
    "            if len(reaction.genes) == 0 and reaction.id not in universals:\n",
    "                obj_coef[reaction.reverse_variable] = 1\n",
    "                obj_coef[reaction.forward_variable] = 1\n",
    "                gfreactions.append(reaction)\n",
    "        genome_hash[genomeid][\"Gapfilled reactions\"] = len(gfreactions)\n",
    "        model.objective.set_linear_coefficients(obj_coef)\n",
    "        mdlutl.pkgmgr.getpkg(\"KBaseMediaPkg\").build_package(gmmedia)\n",
    "        solution = model.optimize()\n",
    "        basegf = 0\n",
    "        for rxn in gfreactions:\n",
    "            if abs(solution.fluxes[rxn.id]) < 0.00000001:\n",
    "                basegf += 1\n",
    "        genome_hash[genomeid][\"Base useless gapfill\"] = basegf\n",
    "        for cpdset in compound_sets:\n",
    "            input_dictionary = {}\n",
    "            for cpd in cpdset:\n",
    "                input_dictionary[cpd] = 1\n",
    "            newmedia = MSMedia.from_dict(input_dictionary)\n",
    "            newmedia.merge(gmmedia)\n",
    "            mdlutl.pkgmgr.getpkg(\"KBaseMediaPkg\").build_package(newmedia)\n",
    "            solution = model.optimize()\n",
    "            gapfilling = 0\n",
    "            for rxn in gfreactions:\n",
    "                if abs(solution.fluxes[rxn.id]) < 0.00000001:\n",
    "                    gapfilling += 1\n",
    "            genome_hash[genomeid][\"auxo\"][\",\".join(cpdset)] = gapfilling-genome_hash[genomeid][\"Base useless gapfill\"]\n",
    "        genome_hash[genomeid][\"reactions\"] = len(model.reactions)\n",
    "        genome_hash[genomeid][\"mdlgenes\"] = len(model.genes)\n",
    "        with open('auxo_hash.json', 'w') as outfile:\n",
    "            json.dump(genome_hash, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'compound_sets' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-fc9b72a506e0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mauxotrophy_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"GenomeID\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"Reactions\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"MdlGenes\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"Useless gapfill\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"Original gapfill\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mcpdset\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcompound_sets\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mcpdstring\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\",\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcpdset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mauxotrophy_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcpdstring\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'compound_sets' is not defined"
     ]
    }
   ],
   "source": [
    "auxotrophy_data = {\"GenomeID\":[],\"Reactions\":[],\"MdlGenes\":[],\"Useless gapfill\":[],\"Original gapfill\":[]}\n",
    "for cpdset in compound_sets:\n",
    "    cpdstring = \",\".join(cpdset)\n",
    "    auxotrophy_data[cpdstring] = []\n",
    "items = []\n",
    "with open('genome_list.json') as json_file:\n",
    "    items = json.load(json_file)\n",
    "for item in items:\n",
    "    item = item+\".RAST\"\n",
    "    if item in genome_hash:\n",
    "        data = genome_hash[item]\n",
    "        auxotrophy_data[\"GenomeID\"].append(item)\n",
    "        if \"Gapfilled reactions\" in data:\n",
    "            auxotrophy_data[\"Original gapfill\"].append(data[\"Gapfilled reactions\"])\n",
    "        else:\n",
    "            auxotrophy_data[\"Original gapfill\"].append(\"\")\n",
    "        if \"Base useless gapfill\" in data:\n",
    "            auxotrophy_data[\"Useless gapfill\"].append(data[\"Base useless gapfill\"])\n",
    "        else:\n",
    "            auxotrophy_data[\"Useless gapfill\"].append(\"\")\n",
    "        if \"reactions\" in data:\n",
    "            auxotrophy_data[\"Reactions\"].append(data[\"reactions\"])\n",
    "        else:\n",
    "            auxotrophy_data[\"Reactions\"].append(\"\")\n",
    "        if \"mdlgenes\" in data:\n",
    "            auxotrophy_data[\"MdlGenes\"].append(data[\"mdlgenes\"])\n",
    "        else:\n",
    "            auxotrophy_data[\"MdlGenes\"].append(\"\")  \n",
    "        if \"auxo\" in data:\n",
    "            for cpdset in compound_sets:\n",
    "                cpdstring = \",\".join(cpdset)\n",
    "                if cpdstring in data[\"auxo\"]:\n",
    "                    auxotrophy_data[cpdstring].append(data[\"auxo\"][cpdstring])\n",
    "                else:\n",
    "                    auxotrophy_data[cpdstring].append(\"\")\n",
    "        else:\n",
    "            for cpdset in compound_sets:\n",
    "                cpdstring = \",\".join(cpdset)\n",
    "                auxotrophy_data[cpdstring].append(\"\")\n",
    "df = pd.DataFrame(auxotrophy_data)\n",
    "df.to_csv(\"AuxotrophyData.csv\")\n",
    "HTML(df.to_html(render_links=True, escape=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genome_hash' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-50377da5ed5d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mitems\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mitem\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\".RAST\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genome_hash' is not defined"
     ]
    }
   ],
   "source": [
    "#Printing dataframe\n",
    "data = {\"GenomeID\":[],\"Taxonomy\":[],\"Species\":[],\"Genes\":[],\"MdlGenes\":[],\"MdlReactions\":[],\n",
    "        \"GFReactions\":[],\"Blocked\":[],\"OldReactions\":[],\"OldGenes\":[],\"OldGFReactions\":[]}\n",
    "\n",
    "items = []\n",
    "with open('genome_list.json') as json_file:\n",
    "    items = json.load(json_file)\n",
    "\n",
    "for item in items:\n",
    "    item = item+\".RAST\"\n",
    "    if item not in genome_hash:\n",
    "        print(item)\n",
    "    if item in genome_hash:\n",
    "        gdata = genome_hash[item]\n",
    "        data[\"GenomeID\"].append(item)\n",
    "        data[\"Genes\"].append(gdata[\"genes\"])\n",
    "        data[\"Taxonomy\"].append(gdata[\"taxonomy\"])\n",
    "        data[\"Species\"].append(gdata[\"species\"])\n",
    "        if \"mdlgenes\" in gdata:\n",
    "            data[\"MdlGenes\"].append(gdata[\"mdlgenes\"])\n",
    "            data[\"MdlReactions\"].append(gdata[\"reactions\"])\n",
    "            data[\"GFReactions\"].append(gdata[\"gfreactions\"])\n",
    "            data[\"Blocked\"].append(gdata[\"blocked\"])\n",
    "        else:\n",
    "            data[\"MdlGenes\"].append(0)\n",
    "            data[\"MdlReactions\"].append(0)\n",
    "            data[\"GFReactions\"].append(0)\n",
    "            data[\"Blocked\"].append(0)\n",
    "        if \"old_mdlgenes\" in gdata:\n",
    "            data[\"OldReactions\"].append(gdata[\"old_reactions\"])\n",
    "            data[\"OldGenes\"].append(gdata[\"old_mdlgenes\"])\n",
    "            data[\"OldGFReactions\"].append(gdata[\"old_gfreactions\"])\n",
    "        else:\n",
    "            data[\"OldReactions\"].append(0)\n",
    "            data[\"OldGenes\"].append(0)\n",
    "            data[\"OldGFReactions\"].append(0)\n",
    "df = pd.DataFrame(data)\n",
    "df.to_csv(\"AuxotrophyDashboard.csv\")\n",
    "HTML(df.to_html(render_links=True, escape=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genome_hash' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-50377da5ed5d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mitems\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mitem\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\".RAST\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genome_hash' is not defined"
     ]
    }
   ],
   "source": [
    "#Printing dataframe\n",
    "data = {\"GenomeID\":[],\"Taxonomy\":[],\"Species\":[],\"Genes\":[],\"MdlGenes\":[],\"MdlReactions\":[],\n",
    "        \"GFReactions\":[],\"Blocked\":[],\"OldReactions\":[],\"OldGenes\":[],\"OldGFReactions\":[]}\n",
    "\n",
    "items = []\n",
    "with open('genome_list.json') as json_file:\n",
    "    items = json.load(json_file)\n",
    "\n",
    "for item in items:\n",
    "    item = item+\".RAST\"\n",
    "    if item not in genome_hash:\n",
    "        print(item)\n",
    "    if item in genome_hash:\n",
    "        gdata = genome_hash[item]\n",
    "        data[\"GenomeID\"].append(item)\n",
    "        data[\"Genes\"].append(gdata[\"genes\"])\n",
    "        data[\"Taxonomy\"].append(gdata[\"taxonomy\"])\n",
    "        data[\"Species\"].append(gdata[\"species\"])\n",
    "        if \"mdlgenes\" in gdata:\n",
    "            data[\"MdlGenes\"].append(gdata[\"mdlgenes\"])\n",
    "            data[\"MdlReactions\"].append(gdata[\"reactions\"])\n",
    "            data[\"GFReactions\"].append(gdata[\"gfreactions\"])\n",
    "            data[\"Blocked\"].append(gdata[\"blocked\"])\n",
    "        else:\n",
    "            data[\"MdlGenes\"].append(0)\n",
    "            data[\"MdlReactions\"].append(0)\n",
    "            data[\"GFReactions\"].append(0)\n",
    "            data[\"Blocked\"].append(0)\n",
    "        if \"old_mdlgenes\" in gdata:\n",
    "            data[\"OldReactions\"].append(gdata[\"old_reactions\"])\n",
    "            data[\"OldGenes\"].append(gdata[\"old_mdlgenes\"])\n",
    "            data[\"OldGFReactions\"].append(gdata[\"old_gfreactions\"])\n",
    "        else:\n",
    "            data[\"OldReactions\"].append(0)\n",
    "            data[\"OldGenes\"].append(0)\n",
    "            data[\"OldGFReactions\"].append(0)\n",
    "df = pd.DataFrame(data)\n",
    "df.to_csv(\"AuxotrophyDashboard.csv\")\n",
    "HTML(df.to_html(render_links=True, escape=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genome_hash' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-50377da5ed5d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mitems\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mitem\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\".RAST\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genome_hash' is not defined"
     ]
    }
   ],
   "source": [
    "#Printing dataframe\n",
    "data = {\"GenomeID\":[],\"Taxonomy\":[],\"Species\":[],\"Genes\":[],\"MdlGenes\":[],\"MdlReactions\":[],\n",
    "        \"GFReactions\":[],\"Blocked\":[],\"OldReactions\":[],\"OldGenes\":[],\"OldGFReactions\":[]}\n",
    "\n",
    "items = []\n",
    "with open('genome_list.json') as json_file:\n",
    "    items = json.load(json_file)\n",
    "\n",
    "for item in items:\n",
    "    item = item+\".RAST\"\n",
    "    if item not in genome_hash:\n",
    "        print(item)\n",
    "    if item in genome_hash:\n",
    "        gdata = genome_hash[item]\n",
    "        data[\"GenomeID\"].append(item)\n",
    "        data[\"Genes\"].append(gdata[\"genes\"])\n",
    "        data[\"Taxonomy\"].append(gdata[\"taxonomy\"])\n",
    "        data[\"Species\"].append(gdata[\"species\"])\n",
    "        if \"mdlgenes\" in gdata:\n",
    "            data[\"MdlGenes\"].append(gdata[\"mdlgenes\"])\n",
    "            data[\"MdlReactions\"].append(gdata[\"reactions\"])\n",
    "            data[\"GFReactions\"].append(gdata[\"gfreactions\"])\n",
    "            data[\"Blocked\"].append(gdata[\"blocked\"])\n",
    "        else:\n",
    "            data[\"MdlGenes\"].append(0)\n",
    "            data[\"MdlReactions\"].append(0)\n",
    "            data[\"GFReactions\"].append(0)\n",
    "            data[\"Blocked\"].append(0)\n",
    "        if \"old_mdlgenes\" in gdata:\n",
    "            data[\"OldReactions\"].append(gdata[\"old_reactions\"])\n",
    "            data[\"OldGenes\"].append(gdata[\"old_mdlgenes\"])\n",
    "            data[\"OldGFReactions\"].append(gdata[\"old_gfreactions\"])\n",
    "        else:\n",
    "            data[\"OldReactions\"].append(0)\n",
    "            data[\"OldGenes\"].append(0)\n",
    "            data[\"OldGFReactions\"].append(0)\n",
    "df = pd.DataFrame(data)\n",
    "df.to_csv(\"AuxotrophyDashboard.csv\")\n",
    "HTML(df.to_html(render_links=True, escape=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genome_hash' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-64801bda218c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mitems\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mitem\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\".RAST\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfilter_hash\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genome_hash' is not defined"
     ]
    }
   ],
   "source": [
    "#Printing dataframe\n",
    "data = {\"GenomeID\":[],\"Taxonomy\":[],\"Species\":[],\"Genes\":[],\"MdlGenes\":[],\"MdlReactions\":[],\n",
    "        \"GFReactions\":[],\"Blocked\":[],\"OldReactions\":[],\"OldGenes\":[],\"OldGFReactions\":[]}\n",
    "\n",
    "items = []\n",
    "with open('genome_list.json') as json_file:\n",
    "    items = json.load(json_file)\n",
    "\n",
    "for item in items:\n",
    "    item = item+\".RAST\"\n",
    "    if item not in genome_hash:\n",
    "        print(item)\n",
    "    if item in filter_hash:\n",
    "        gdata = genome_hash[item]\n",
    "        data[\"GenomeID\"].append(item)\n",
    "        data[\"Genes\"].append(gdata[\"genes\"])\n",
    "        data[\"Taxonomy\"].append(gdata[\"taxonomy\"])\n",
    "        data[\"Species\"].append(gdata[\"species\"])\n",
    "        if \"mdlgenes\" in gdata:\n",
    "            data[\"MdlGenes\"].append(gdata[\"mdlgenes\"])\n",
    "            data[\"MdlReactions\"].append(gdata[\"reactions\"])\n",
    "            data[\"GFReactions\"].append(gdata[\"gfreactions\"])\n",
    "            data[\"Blocked\"].append(gdata[\"blocked\"])\n",
    "        else:\n",
    "            data[\"MdlGenes\"].append(0)\n",
    "            data[\"MdlReactions\"].append(0)\n",
    "            data[\"GFReactions\"].append(0)\n",
    "            data[\"Blocked\"].append(0)\n",
    "        if \"old_mdlgenes\" in gdata:\n",
    "            data[\"OldReactions\"].append(gdata[\"old_reactions\"])\n",
    "            data[\"OldGenes\"].append(gdata[\"old_mdlgenes\"])\n",
    "            data[\"OldGFReactions\"].append(gdata[\"old_gfreactions\"])\n",
    "        else:\n",
    "            data[\"OldReactions\"].append(0)\n",
    "            data[\"OldGenes\"].append(0)\n",
    "            data[\"OldGFReactions\"].append(0)\n",
    "df = pd.DataFrame(data)\n",
    "df.to_csv(\"ProblemAuxoDashboard.csv\")\n",
    "HTML(df.to_html(render_links=True, escape=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_list = "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
