{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ac88d15d-764e-477d-ba9d-fd13408d0edd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cobrakbase 0.3.1\n"
     ]
    }
   ],
   "source": [
    "# import kbase\n",
    "import os\n",
    "local_cobrakbase_path = os.path.join('C:', 'Users', 'Andrew Freiburger','Documents','Argonne','cobrakbase')\n",
    "os.environ[\"HOME\"] = local_cobrakbase_path\n",
    "import cobrakbase\n",
    "with open(\"D:/Users/Andrew Freiburger/Documents/Argonne/kbase_token.txt\") as token_file:\n",
    "    kbase_api = cobrakbase.KBaseAPI(token_file.readline())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7fc1aac4-bcbd-4985-b49f-41e57242a9e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "F_QMMZ011 = kbase_api.get_from_ws(\"146461/195/1\") ; F_LQBF011 = kbase_api.get_from_ws(\"146461/194/1\")\n",
    "F_MAXM011 = kbase_api.get_from_ws(\"146461/193/1\") ; F_QNAY011 = kbase_api.get_from_ws(\"146461/192/1\")\n",
    "F_PQXE011 = kbase_api.get_from_ws(\"146461/191/1\") ; F_PQXD011 = kbase_api.get_from_ws(\"146461/190/1\")\n",
    "F_MANV011 = kbase_api.get_from_ws(\"146461/189/1\")\n",
    "print(\"done 1/4\", end=\"\\r\")\n",
    "F_JAJSZR000000000 = kbase_api.get_from_ws(\"146461/180/1\") ; F_JAJSZV000000000 = kbase_api.get_from_ws(\"146461/181/1\")\n",
    "F_JAJSZY000000000 = kbase_api.get_from_ws(\"146461/179/1\") ; F_JAJSZU000000000 = kbase_api.get_from_ws(\"146461/178/1\")\n",
    "F_JAJSZW000000000 = kbase_api.get_from_ws(\"146461/177/1\") ; F_JAJSZT000000000 = kbase_api.get_from_ws(\"146461/176/1\")\n",
    "F_JAJSZQ000000000 = kbase_api.get_from_ws(\"146461/174/1\") ; F_JAJSZP000000000 = kbase_api.get_from_ws(\"146461/172/1\")\n",
    "F_JAJSZO000000000 = kbase_api.get_from_ws(\"146461/170/1\")\n",
    "print(\"done 2/4\", end=\"\\r\")\n",
    "F_IAIORU011 = kbase_api.get_from_ws(\"146461/168/1\") ; F_IAIORM011 = kbase_api.get_from_ws(\"146461/166/1\") ; F_IAFDLJ011 = kbase_api.get_from_ws(\"146461/164/1\") ; F_IAFDLI011 = kbase_api.get_from_ws(\"146461/162/1\")\n",
    "F_IAFDKW011 = kbase_api.get_from_ws(\"146461/160/1\") ; F_IAFDJY011 = kbase_api.get_from_ws(\"146461/158/1\") ; F_IAFDIU011 = kbase_api.get_from_ws(\"146461/156/1\") ; F_IAFDIR011 = kbase_api.get_from_ws(\"146461/154/1\")\n",
    "F_IAFDIO011 = kbase_api.get_from_ws(\"146461/152/1\") ; F_IAFDIM011 = kbase_api.get_from_ws(\"146461/150/1\") ; F_IAFDIH011 = kbase_api.get_from_ws(\"146461/148/1\") ; F_IAFDIE011 = kbase_api.get_from_ws(\"146461/146/1\")\n",
    "F_IAFDGV011 = kbase_api.get_from_ws(\"146461/144/1\") ; F_IAFDGK011 = kbase_api.get_from_ws(\"146461/142/1\") ; F_IAFDGC011 = kbase_api.get_from_ws(\"146461/140/1\") ; F_IAFDFV011 = kbase_api.get_from_ws(\"146461/138/1\")\n",
    "F_IAFDFS011 = kbase_api.get_from_ws(\"146461/136/1\") ; F_IAFCZG011 = kbase_api.get_from_ws(\"146461/134/1\") ; F_JAABVH011 = kbase_api.get_from_ws(\"146461/131/1\") ; F_JAABVG011 = kbase_api.get_from_ws(\"146461/128/1\")\n",
    "F_JAAXOL011 = kbase_api.get_from_ws(\"146461/127/1\") ; F_JABZFQ011 = kbase_api.get_from_ws(\"146461/126/1\") ; F_CAJIMK011 = kbase_api.get_from_ws(\"146461/124/1\") ; F_CAJIMJ011 = kbase_api.get_from_ws(\"146461/122/1\")\n",
    "print(\"done 3/4\", end=\"\\r\")\n",
    "JAAXOL011 = kbase_api.get_from_ws(\"146461/114/1\") ; PQXD011 = kbase_api.get_from_ws(\"146461/115/1\") ; JAABVG011 = kbase_api.get_from_ws(\"146461/111/1\") ; PQXE011 = kbase_api.get_from_ws(\"146461/110/1\")\n",
    "CAJIMJ011 = kbase_api.get_from_ws(\"146461/111/1\") ; CAJIMK011 = kbase_api.get_from_ws(\"146461/105/1\")\n",
    "print(\"done 4/4\", end=\"\\r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cc43b685-c15c-4997-8708-84de16ce63d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes = [F_QMMZ011, F_LQBF011, F_MAXM011, F_QNAY011, F_PQXE011, F_PQXD011, F_MANV011, \n",
    "           F_JAJSZR000000000, F_JAJSZV000000000, F_JAJSZY000000000, F_JAJSZU000000000, F_JAJSZW000000000, F_JAJSZT000000000, F_JAJSZQ000000000, F_JAJSZP000000000, F_JAJSZO000000000, \n",
    "           F_IAIORU011, F_IAIORM011, F_IAFDLJ011, F_IAFDLI011, F_IAFDKW011, F_IAFDJY011, F_IAFDIU011, F_IAFDIR011, F_IAFDIO011, F_IAFDIM011, F_IAFDIH011, F_IAFDIE011, \n",
    "           F_IAFDGV011, F_IAFDGK011, F_IAFDGC011, F_IAFDFV011, F_IAFDFS011, F_IAFCZG011, F_JAABVH011, F_JAABVG011, F_JAAXOL011, F_JABZFQ011, F_CAJIMK011, F_CAJIMJ011, \n",
    "           JAAXOL011, PQXD011, JAABVG011, PQXE011, CAJIMJ011, CAJIMK011]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a00f2962-bbf2-4b3c-827c-09b390e619c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['OBJECT_TYPE', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_from_json', '_repr_html_', '_to_json', '_to_object', 'add_cdss', 'add_features', 'alias_hash', 'assembly_ref', 'cdss', 'copied', 'copy_source_inaccessible', 'created', 'creator', 'data', 'data_keys', 'dna_size', 'domain', 'epoch', 'exclude_dict', 'feature_counts', 'features', 'from_dna_fasta', 'from_fasta', 'from_kbase_data', 'from_protein_sequences_hash', 'gc_content', 'genetic_code', 'genome_tiers', 'get_data', 'get_kbase_args', 'id', 'info', 'md5', 'molecule_type', 'num_contigs', 'orig_wsid', 'path', 'provenance', 'refs', 'scientific_name', 'search_for_gene', 'source', 'taxon_ref', 'taxonomy', 'to_fasta', 'to_kbase_data']\n",
      "QMMZ01.1.gbff_genome\n",
      "['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_genome', '_repr_html_', 'add_ontology_term', 'aliases', 'cdss', 'description', 'dna_sequence', 'dna_sequence_length', 'extract_functions', 'from_kbase_data', 'functions', 'id', 'location', 'md5', 'ontology_terms', 'protein_translation', 'protein_translation_length', 'seq', 'split_annotation', 'split_function', 'to_kbase_data']\n",
      "DRH21_00005\n",
      "[['locus_tag', 'DRH21_00005'], ['protein_id', 'RLC27111.1']]\n",
      "['DRH21_00005_CDS_1']\n",
      "['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_genome', 'add_ontology_term', 'aliases', 'description', 'dna_md5', 'dna_sequence', 'dna_sequence_length', 'from_kbase_data', 'id', 'location', 'ontology_terms', 'protein_md5', 'protein_translation_length', 'seq', 'to_kbase_data']\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "print(dir(genomes[0]))\n",
    "print(genomes[0].id)\n",
    "print(dir(genomes[0].features[0]))\n",
    "print(genomes[0].features[0])\n",
    "print(genomes[0].features[0].aliases)\n",
    "print(genomes[0].features[0].cdss)\n",
    "print(dir(genomes[0].cdss[0]))\n",
    "print(genomes[0].cdss[0].aliases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d404a36d-ee0b-4812-b29a-5bed15137b17",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# from pandas import DataFrame\n",
    "genome_genes_mapping = {}\n",
    "for genome in genomes:\n",
    "    genome_genes_mapping[genome.id] = []\n",
    "    for feature in genome.features:\n",
    "        for alias in feature.aliases:\n",
    "            genome_genes_mapping[genome.id].append(alias[1])\n",
    "genes_genome_mapping = {gene:genome for genome, genes in genome_genes_mapping.items() for gene in genes}\n",
    "# print(genome_genes_mapping[genomes[0].id])\n",
    "# print(len(genes_genome_mapping))\n",
    "# print({k: len(v) for k,v in gene_genome_mapping.items()})\n",
    "# df = DataFrame(gene_genome_mapping)\n",
    "# display(df)\n",
    "\n",
    "from json import dump\n",
    "with open(\"gene_genome_mapping.json\", \"w\") as out:  dump(genes_genome_mapping, out, indent=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "0918ec8f-f441-49be-9e3f-8cf94c26a759",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_genes(geneID):\n",
    "    if geneID not in genes_genome_mapping:  return \" \"\n",
    "    return genes_genome_mapping[geneID]\n",
    "\n",
    "from pandas import read_csv, concat\n",
    "\n",
    "total_table = read_csv(\"total_GeneID_function.csv\")\n",
    "table5 = read_csv(\"Table_5.csv\")\n",
    "table8 = read_csv(\"Table_8.csv\")\n",
    "combined_table = concat([total_table, table5, table8], ignore_index=True)\n",
    "\n",
    "combined_table[\"KBase_genomeID\"] = combined_table[\"GeneID\"].apply(map_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "19e1543b-a0b4-4fe1-859c-2244fb151956",
   "metadata": {},
   "outputs": [],
   "source": [
    "from numpy import nan\n",
    "mapped_genes = combined_table[combined_table[\"KBase_genomeID\"] != \" \"] # .to_csv(\"KBase_mapped_genomes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "3a4d00d5-32a4-4d52-9e57-ab7409213200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Function</th>\n",
       "      <th>OrganismID</th>\n",
       "      <th>KBase_genomeID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>HmcA</td>\n",
       "      <td>Desulfobacterales_sp._PC51MH44</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>OEU67172.1 (atypical)</td>\n",
       "      <td>HmcA</td>\n",
       "      <td>Desulfobacterales_sp._S5133MH16</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>HmcA</td>\n",
       "      <td>Desulfosarcina_sp._BuS5</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2512526948</td>\n",
       "      <td>HmcA</td>\n",
       "      <td>Desulfococcus_biacutus_KMRActS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2502442176</td>\n",
       "      <td>HmcA</td>\n",
       "      <td>Desulfonema_limicola_Jadebusen_DSM_2076</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3995</th>\n",
       "      <td>1425</td>\n",
       "      <td>0</td>\n",
       "      <td>bzaE</td>\n",
       "      <td>Seep-SRB1g_(str._C00003104)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3996</th>\n",
       "      <td>1426</td>\n",
       "      <td>0</td>\n",
       "      <td>bzaE</td>\n",
       "      <td>Seep-SRB1g_(str._C00003106)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3997</th>\n",
       "      <td>1427</td>\n",
       "      <td>0</td>\n",
       "      <td>bzaE</td>\n",
       "      <td>Seep-SRB1g_(str._CR10073A)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3998</th>\n",
       "      <td>1428</td>\n",
       "      <td>0</td>\n",
       "      <td>bzaE</td>\n",
       "      <td>Seep-SRB1g_(str._CR10073B)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3999</th>\n",
       "      <td>1429</td>\n",
       "      <td>0</td>\n",
       "      <td>bzaE</td>\n",
       "      <td>Seep-SRB1g_(str._S7142MS4)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3644 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0                 GeneID Function  \\\n",
       "0              0                      -     HmcA   \n",
       "1              1  OEU67172.1 (atypical)     HmcA   \n",
       "2              2                      -     HmcA   \n",
       "3              3             2512526948     HmcA   \n",
       "4              4             2502442176     HmcA   \n",
       "...          ...                    ...      ...   \n",
       "3995        1425                      0     bzaE   \n",
       "3996        1426                      0     bzaE   \n",
       "3997        1427                      0     bzaE   \n",
       "3998        1428                      0     bzaE   \n",
       "3999        1429                      0     bzaE   \n",
       "\n",
       "                                   OrganismID KBase_genomeID  \n",
       "0              Desulfobacterales_sp._PC51MH44                 \n",
       "1             Desulfobacterales_sp._S5133MH16                 \n",
       "2                     Desulfosarcina_sp._BuS5                 \n",
       "3              Desulfococcus_biacutus_KMRActS                 \n",
       "4     Desulfonema_limicola_Jadebusen_DSM_2076                 \n",
       "...                                       ...            ...  \n",
       "3995              Seep-SRB1g_(str._C00003104)                 \n",
       "3996              Seep-SRB1g_(str._C00003106)                 \n",
       "3997               Seep-SRB1g_(str._CR10073A)                 \n",
       "3998               Seep-SRB1g_(str._CR10073B)                 \n",
       "3999               Seep-SRB1g_(str._S7142MS4)                 \n",
       "\n",
       "[3644 rows x 5 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO clean and filter the DataFrame that is searched for geneIDs\n",
    "# TODO export the missing geneIDs\n",
    "combined_table[combined_table[\"KBase_genomeID\"] == \" \"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4e63e303-b2ba-4c17-b613-75fe5445e32e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({('Desulfococcus_sp._C00003106', 'MAXM01.1.gbff_genome'): 4,\n",
       "         ('Desulfococcus_sp._C00003104', 'MANV01.1.gbff_genome'): 3,\n",
       "         ('Seep-SRB1g_sp001751015_a', 'MANV01.1.gbff_genome'): 3,\n",
       "         ('Seep-SRB1a_sp014237365', 'JAAXOL01.1.gbff_genome'): 4,\n",
       "         ('Seep-SRB1a_sp014237365a', 'JABZFQ01.1.gbff_genome'): 4,\n",
       "         ('Seep-SRB1g_sp001751015_b', 'MAXM01.1.gbff_genome'): 3,\n",
       "         ('E20 SRB2', 'PQXE01.1.gbff_genome'): 2,\n",
       "         ('G37 SRB2', 'PQXD01.1.gbff_genome'): 2,\n",
       "         ('Seep-SRB2_Meg22_02_Bin_69', 'JAFDFS01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB2_Meg22_1012_Bin_255', 'JAFDGK01.1.gbff_genome'): 2,\n",
       "         ('Seep-SRB2_Meg22_1214_Bin_80', 'JAFDIH01.1.gbff_genome'): 2,\n",
       "         ('Seep-SRB2_Meg22_1416_Bin_56', 'JAFDIO01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB2_Meg22_1618_Bin_165', 'JAFDIU01.1.gbff_genome'): 2,\n",
       "         ('SeepSRB2_G37_sp003194485', 'PQXD01.1.gbff_genome'): 1,\n",
       "         ('SeepSRB2_sp._PB10259FWG173', 'JAJSZQ000000000.gbff_genome'): 1,\n",
       "         ('SeepSRB2_sp._PBNA091R45FWG174', 'JAJSZR000000000.gbff_genome'): 1,\n",
       "         ('SeepSRB2_sp003194495', 'PQXE01.1.gbff_genome'): 1,\n",
       "         ('SeepSRB2_sp00364560', 'QNAY01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a_AB_03_Bin_172', 'JAFCZG01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a_Meg22_02_Bin_90', 'JAFDFV01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a_Meg22_24_Bin_68', 'JAFDJY01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a_Meg22_46_Bin_236', 'JAFDKW01.1.gbff_genome'): 1,\n",
       "         ('SeepSRB1a_sp._S7142MS3', 'JAABVG01.1.gbff_genome'): 1,\n",
       "         ('SeepSRB1a_sp._SM7059A', 'JAJSZW000000000.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a sp. SM7059A', 'JAJSZW000000000.gbff_genome'): 32,\n",
       "         ('Seep-SRB1a sp. CR9063A', 'JAJSZT000000000.gbff_genome'): 40,\n",
       "         ('Seep-SRB2 sp. G37', 'PQXD01.1.gbff_genome'): 21,\n",
       "         ('E20 Seep-SRB2', 'PQXE01.1.gbff_genome'): 33,\n",
       "         ('Seep-SRB2_sp._PB10259FWG173', 'JAJSZQ000000000.gbff_genome'): 23,\n",
       "         ('Seep-SRB2_sp._PBNA091R45FWG174', 'JAJSZR000000000.gbff_genome'): 28,\n",
       "         (nan, 'MAXM01.1.gbff_genome'): 5,\n",
       "         (nan, 'JAFCZG01.1.gbff_genome'): 7,\n",
       "         (nan, 'JAAXOL01.1.gbff_genome'): 4,\n",
       "         (nan, 'JAFDFV01.1.gbff_genome'): 2,\n",
       "         (nan, 'JAFDIR01.1.gbff_genome'): 2,\n",
       "         (nan, 'JAABVH01.1.gbff_genome'): 1,\n",
       "         (nan, 'JAABVG01.1.gbff_genome'): 2,\n",
       "         (nan, 'MANV01.1.gbff_genome'): 1,\n",
       "         (nan, 'PQXD01.1.gbff_genome'): 1,\n",
       "         (nan, 'JAFDIU01.1.gbff_genome'): 1,\n",
       "         (nan, 'JAFDIH01.1.gbff_genome'): 1,\n",
       "         (nan, 'JAFDIE01.1.gbff_genome'): 1,\n",
       "         (nan, 'PQXE01.1.gbff_genome'): 1,\n",
       "         (nan, 'QNAY01.1.gbff_genome'): 2,\n",
       "         (nan, 'JAJSZY000000000.gbff_genome'): 1,\n",
       "         (nan, 'JAJSZP000000000.gbff_genome'): 1,\n",
       "         (nan, 'JABZFQ01.1.gbff_genome'): 1,\n",
       "         (nan, 'JAFDJY01.1.gbff_genome'): 1,\n",
       "         (nan, 'JAJSZW000000000.gbff_genome'): 1,\n",
       "         (nan, 'JAJSZT000000000.gbff_genome'): 1,\n",
       "         (nan, 'JAJSZU000000000.gbff_genome'): 1,\n",
       "         (nan, 'JAJSZO000000000.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a_sp._S7142MS3', 'JAABVG01.1.gbff_genome'): 3,\n",
       "         ('Seep-SRB1a_sp._Meg22_46_Bin_236', 'JAFDKW01.1.gbff_genome'): 6,\n",
       "         ('Seep-SRB1a_sp._AB_03_Bin_172', 'JAFCZG01.1.gbff_genome'): 10,\n",
       "         ('Seep-SRB2_E20_sp003194495', 'PQXE01.1.gbff_genome'): 10,\n",
       "         ('Seep-SRB2_G37_sp003194485', 'PQXD01.1.gbff_genome'): 28,\n",
       "         ('Seep-SRB2 sp. Meg22_1214_Bin_80', 'JAFDIH01.1.gbff_genome'): 9,\n",
       "         ('Seep-SRB1g_sp001751015_a_DIET_cluster.txt',\n",
       "          'MANV01.1.gbff_genome'): 13,\n",
       "         ('Seep-SRB2_sp._4_(str._E20)', 'PQXE01.1.gbff_genome'): 2,\n",
       "         ('Seep-SRB2_sp._5_(str._003645605)', 'QNAY01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB2_sp._1_(str._G37)', 'PQXD01.1.gbff_genome'): 11,\n",
       "         ('Seep-SRB1a_sp._3_(str._014237195)', 'JAAXOL01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a_sp._3_(str._014237365)', 'JABZFQ01.1.gbff_genome'): 1,\n",
       "         ('Seep-SRB1a_sp._4_(str._FWG172)', 'JAJSZP000000000.gbff_genome'): 1})"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "from numpy import array\n",
    "\n",
    "genomeID_mappings = list(zip(mapped_genes[\"OrganismID\"].tolist(), mapped_genes[\"KBase_genomeID\"].tolist()))\n",
    "Counter(genomeID_mappings)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
